### PR TITLE
[CompilerPlugin] Updates after Foundation removal

### PR DESCRIPTION
**Pull request by rintaro**
* Merge `_AtomicBool` into `_SwiftSyntaxCShims`
* Make `_AtomicBool` header only
* Separate `errno` from `_stdio.h`
* Give `swift_name` attributes to C shims

Also, move `SwiftCompilerPlugin.PluginMessageConnection` to `SwiftCompilerPluginMessageHandling`, so other components e.g. `swift-plugin-server` can use it as well.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test Windows

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test Windows

File: Sources/SwiftSyntax/SyntaxArena.swift [hunk 1]
```diff
@@ -24,7 +24,7 @@ fileprivate struct AtomicBool {
   }
 }
 #else
-import _AtomicBool
+import _SwiftSyntaxCShims
 #endif
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxArena.swift:28]: If we’ve got `AtomicBool` available in the CMake build as well now (I never thought of making it header-only, nice idea), I think we can also remove the hack of defining a fake `AtomicBool` when `SWIFT_SYNTAX_BUILD_USING_CMAKE` is sent and instead always use `_SwiftSyntaxCShims`.

**Review by ahoppen — APPROVED** [review]

**Comment by rintaro** [comment]
https://github.com/apple/swift/pull/73224
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/swift/pull/73224
@swift-ci Please test Windows

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test Windows

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test Windows

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test Windows

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/swift-package-manager/pull/7503
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/swift-package-manager/pull/7503
@swift-ci Please test

**Comment by rintaro** [comment]
https://github.com/apple/swift-package-manager/pull/7503
@swift-ci Please test Windows

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test Windows

**Comment by rintaro** [comment]
Leaving the AtomicBool hack as-is for now, because making SwiftSyntax depend on CShims is too much hassle.

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Pull request merged**